manticore/wasm/structure.py

```diff
@@ -0,0 +1,1703 @@
+
+    TODO: We should implement some kind of symbolic memory model
+
+    https://www.w3.org/TR/wasm-core-1/#memory-instances%E2%91%A0
+    """
+
+    data: typing.List[int]  #: The backing array for this memory
+    max: typing.Optional[U32]  #: Optional maximum number of pages the memory can contain
+
+    def grow(self, n: int) -> bool:
+        """
+        Adds n blank pages to the current memory
+
+        See: https://www.w3.org/TR/wasm-core-1/#grow-mem
+
+        :param n: The number of pages to attempt to add
+        :return: True if the operation succeeded, otherwise False
+        """
+        assert len(self.data) % 65536 == 0
+        ln = n + (len(self.data) // 65536)
+        if ln > (2 ** 16):
+            return False
+        if self.max is not None:
+            if ln > self.max:
+                return False
+        self.data.extend(0x0 for _ in range(n * 65536))  # TODO - these should also be symbolic
+        return True
+
+
+@dataclass
+class GlobalInst:
+    """
+    Instance of a global variable. Stores the value (calculated from evaluating a Global.init) and the mutable flag
+    (taken from GlobalType.mut)
+
+    https://www.w3.org/TR/wasm-core-1/#global-instances%E2%91%A0
+    """
+
+    value: Value  #: The actual value of this global
+    mut: bool  #: Whether the global can be modified
+
+
+@dataclass
+class ExportInst:
+    """
+    Runtime representation of any thing that can be exported
+
+    https://www.w3.org/TR/wasm-core-1/#export-instances%E2%91%A0
+    """
+
+    name: Name  #: The name to export under
+    value: ExternVal  #: FuncAddr, TableAddr, MemAddr, or GlobalAddr
+
+
+class Store:
+    """
+    Implementation of the WASM store. Nothing fancy here, just collects lists of functions, tables, memories, and
+    globals. Because the store is not atomic, instructions SHOULD NOT make changes to the Store or any of its contents
+    (including memories and global variables) before raising a Concretize exception.
+
+    https://www.w3.org/TR/wasm-core-1/#store%E2%91%A0
+    """
+
+    __slots__ = ["funcs", "tables", "mems", "globals"]
+    funcs: typing.List[ProtoFuncInst]
+    tables: typing.List[TableInst]
+    mems: typing.List[MemInst]
+    globals: typing.List[GlobalInst]
+
+    def __init__(self):
+        self.funcs = []
+        self.tables = []
+        self.mems = []
+        self.globals = []
+
+    def __getstate__(self):
+        state = {
+            "funcs": self.funcs,
+            "tables": self.tables,
+            "mems": self.mems,
+            "globals": self.globals,
+        }
+        return state
+
+    def __setstate__(self, state):
+        self.funcs = state["funcs"]
+        self.tables = state["tables"]
+        self.mems = state["mems"]
+        self.globals = state["globals"]
+
+
+def _eval_maybe_symbolic(constraints, expression) -> bool:
+    if issymbolic(expression):
+        return solver.must_be_true(constraints, expression)
+    return True if expression else False
```

Comment:
Could this be written as
```
return expression is not None
```
I'm also not quite sure what we're checking for in `expression`

Reply:
I think this is a typing thing. `expression` isn't explicitly constrained to `typing.Union[Bitvec, bool]` because mypy can't figure out that `issymbolic` will only be true for `BitVec`s, so to confirm that this will always return a bool, I do `return True if expression else False` instead of just `return expression`

Reply:
Got it. That makes sense if `expression` will always be a `bool` at this point and this is just to make mypy happy 👍